allspice/utils/list_components.py

```diff
@@ -172,34 +176,85 @@ def list_components_for_altium(
+    #    after which this is no longer used.
+    # 2. The path of the file in the design reuse repository: Used for fetching
+    #    the JSON, after which this is no longer used.
+    # 3. The name stored in the `filename` property of the SheetRef: This is the
+    #    stem of the path. We need to match a SheetRef to the device sheet it is
+    #    referring to using this, which is why `device_sheet_jsons` uses these
+    #    as the keys.
+
+    for schdoc_file in project_documents:
+        schdoc_path_from_repo_root = _resolve_prjpcb_relative_path(schdoc_file, prjpcb_file)
+
+        schdoc_json = _fetch_generated_json(
             repository,
-            _resolve_prjpcb_relative_path(schdoc_file, prjpcb_file),
+            schdoc_path_from_repo_root,
             ref,
         )
-        for schdoc_file in schdoc_files_in_proj
-    }
-    schdoc_entries = {
-        schdoc_file: [value for value in schdoc_json.values() if isinstance(value, dict)]
-        for schdoc_file, schdoc_json in schdoc_jsons.items()
-    }
-    schdoc_refs = {
-        schdoc_file: [entry for entry in entries if entry.get("type") == "SheetRef"]
-        for schdoc_file, entries in schdoc_entries.items()
-    }
-    independent_sheets, hierarchy = _build_schdoc_hierarchy(schdoc_refs)
+        schdoc_jsons[schdoc_path_from_repo_root] = schdoc_json
+
+    for device_sheet in device_sheets:
+        device_sheet_repo, device_sheet_path = _find_device_sheet(
+            device_sheet,
+            repository,
+            prjpcb_file,
+            design_reuse_repos,
+            allspice_client.logger,
+        )
+        device_sheet_json = _fetch_generated_json(
+            device_sheet_repo,
+            device_sheet_path.as_posix(),
+            # Note the default branch here - we can't assume the same ref is
+            # available.
+            device_sheet_repo.default_branch,
+        )
+        device_sheet_jsons[device_sheet_path.stem] = device_sheet_json
+
+    independent_sheets, hierarchy = _build_schdoc_hierarchy(schdoc_jsons, device_sheet_jsons)
+
+    allspice_client.logger.debug("Independent sheets: %s", independent_sheets)
+    allspice_client.logger.debug("Hierarchy: %s", hierarchy)
+
+    # Now we can build a combined mapping of documents and device sheets:
+    sheets_to_components = {}
+    for schdoc_file, schdoc_json in schdoc_jsons.items():
+        schdoc_components = [
+            value
+            for value in schdoc_json.values()
+            if isinstance(value, dict) and value.get("type") == "Component"
+        ]
+        sheets_to_components[schdoc_file] = schdoc_components
+    for device_sheet_name, device_sheet_json in device_sheet_jsons.items():
+        device_sheet_components = [
+            value
+            for value in device_sheet_json.values()
+            if isinstance(value, dict) and value.get("type") == "Component"
+        ]
+        sheets_to_components[device_sheet_name] = device_sheet_components
```

Comment:
There shouldn't be, because the schdoc files are paths while device sheets are names. Theoretically it is possible but in the same situation Altium would also have difficulty identifying which sheet it is using. Might be worth testing @polypour.

Reply:
@shrik450 comment above is true. Altium identifies device sheets by name, primarily. I've tried including multiple folders with the same device sheets under different directory nestings with the same names and it only picks up the first copy seen.